lab/skel_graph/shortest_path/johnson/main.cpp

```diff
@@ -0,0 +1,226 @@
+        pq.push({d[source], source});
+
+        while (!pq.empty()) {
+            // Extragem nodul cu costul estimat minim fata de sursa
+            auto entry = pq.top();
+            pq.pop();
+
+            int cost = entry.first;
+            int node = entry.second;
+
+            // Daca intrarea curenta nu este up-to-date
+            // Adica s-a modificat costul fata de cel deja existent in heap
+            if (cost > d[node]) {
+                continue;
+            }
+
+            // Pentru fiecare vecin incercam sa relaxam costul de la sursa
+            // la el trecand prin nodul node
+            // for (auto &edge : adj[node]) {
+            for (int i = 1; i <= n; ++i) {
+                int neighbour = i;
+                int cost_edge = adj[node][i];
+
+                // Daca obtinem un cost mai mic trecand prin node
+                if (d[node] + cost_edge < d[neighbour]) {
+                    // Salvam noua distanta
+                    d[neighbour] = d[node] + cost_edge;
+
+                    // Actualizam costul drumului de la sursa la vecin
+                    pq.push({d[neighbour], neighbour});
+                }
+            }
+        }
+    }
+
+    bool BellmanFord(int source, vector<int>& d) {
+        // Coada folosita la algoritmul BellmanFord
+        queue<int> q;
+
+        // used[i] = de cate ori a fost folosit nodul i
+        vector<int> used(n + 2, 0);
+
+        // Presupunem ca nu exista drum de la source la i
+        for (int i = 1; i <= n + 1; ++i) {
+            d[i] = oo;
+        }
+
+        // Distanta source -> source = 0
+        d[source] = 0;
+
+        // Adaugam source in coada
+        q.push(source);
+
+        while (!q.empty()) {
+            int node = q.front();
+            q.pop();
+
+            // Crestem numarul de utilizari al nodului node
+            ++used[node];
+            if (used[node] > n + 1) {
+                return true; // Am gasit ciclu de cost negativ
+            }
+
+            // Pentru fiecare vecin incercam sa relaxam costul de la sursa
+            // la el trecand prin nodul node
+            for (int i = 1; i <= n; ++i) {
+                int neighbour = i;
+                int cost_edge = adj[node][i];
+
+                // Daca obtinem un cost mai mic trecand prin node
+                if (d[node] + cost_edge < d[neighbour]) {
+
+                    // Salvam noua distanta
+                    d[neighbour] = d[node] + cost_edge;
+
+                    // Actualizam costul drumului de la sursa la vecin
+                    q.push(neighbour);
+                }
+            }
+        }
+
+        return false;
+    }
+
+    bool Johnson() {
+        // Cream legaturi de la nodul nou la toate celelalte
+        for (int i = 1; i <= n; ++i) {
+            adj[n + 1][i] = 0;
+        }
+
+        // Initializam vectorul pe care il vom trimite la BellmanFord cu valoarea oo
+        vector<int> vertex_weight(n + 2, oo);
```

Comment:
Whoul you mind to use the hame notations from wikipedia? (h). Also is not weight, it's a distance